gdx/src/com/badlogic/gdx/graphics/GLTexture.java

```diff
@@ -1,222 +1,250 @@
+		if (magFilter != null && (force || this.magFilter != magFilter)) {
+			Gdx.gl.glTexParameterf(glTarget, GL20.GL_TEXTURE_MAG_FILTER, magFilter.getGLEnum());
+			this.magFilter = magFilter;
+		}
+	}
+
+	/** Sets the {@link TextureFilter} for this texture for minification and magnification. This will bind this texture!
+	 * @param minFilter the minification filter
+	 * @param magFilter the magnification filter */
+	public void setFilter (TextureFilter minFilter, TextureFilter magFilter) {
+		this.minFilter = minFilter;
+		this.magFilter = magFilter;
+		bind();
+		Gdx.gl.glTexParameterf(glTarget, GL20.GL_TEXTURE_MIN_FILTER, minFilter.getGLEnum());
+		Gdx.gl.glTexParameterf(glTarget, GL20.GL_TEXTURE_MAG_FILTER, magFilter.getGLEnum());
+	}
+
+	/** Destroys the OpenGL Texture as specified by the glHandle. */
+	protected void delete () {
+		if (glHandle != 0) {
+			Gdx.gl.glDeleteTexture(glHandle);
+			glHandle = 0;
+		}
+	}
+
+	@Override
+	public void dispose () {
+		delete();
+	}
+
+	protected static void uploadImageData (int target, TextureData data) {
+		uploadImageData(target, data, 0);
+	}
+
+	public static void uploadImageData (int target, TextureData data, int miplevel) {
+		if (data == null) {
+			// FIXME: remove texture on target?
+			return;
+		}
+
+		if (!data.isPrepared()) data.prepare();
+
+		final TextureDataType type = data.getType();
+		if (type == TextureDataType.Custom) {
+			data.consumeCustomData(target);
+			return;
+		}
+
+		Pixmap pixmap = data.consumePixmap();
+		boolean disposePixmap = data.disposePixmap();
+		if (data.getFormat() != pixmap.getFormat()) {
+			Pixmap tmp = new Pixmap(pixmap.getWidth(), pixmap.getHeight(), data.getFormat());
+			Blending blend = Pixmap.getBlending();
+			Pixmap.setBlending(Blending.None);
+			tmp.drawPixmap(pixmap, 0, 0, 0, 0, pixmap.getWidth(), pixmap.getHeight());
+			Pixmap.setBlending(blend);
+			if (data.disposePixmap()) {
+				pixmap.dispose();
+			}
+			pixmap = tmp;
+			disposePixmap = true;
+		}
+
+		Gdx.gl.glPixelStorei(GL20.GL_UNPACK_ALIGNMENT, 1);
+		if (data.useMipMaps()) {
+			MipMapGenerator.generateMipMap(target, pixmap, pixmap.getWidth(), pixmap.getHeight());
+		} else {
+			Gdx.gl.glTexImage2D(target, miplevel, pixmap.getGLInternalFormat(), pixmap.getWidth(), pixmap.getHeight(), 0,
+				pixmap.getGLFormat(), pixmap.getGLType(), pixmap.getPixels());
+		}
+		if (disposePixmap) {
+			pixmap.dispose();
+		}
+	}
+
+	/** Returns the current state of mipmaps, as managed by this Texture object. <br>
+	 * Note that this method will usually return incorrect values if mipmaps were created by any means other than
+	 * {@link #generateMipMaps} (e.g. by <code>glGenerateMipmap</code>).
+	 * @return */
+	public boolean hasMipmaps () {
+		return hasMipmaps;
+	}
+
+	/** Idempotent. Reversible only by calling {@link #reload()}. Calls {@link #bind() before actual mipmap generation. */
+	public void generateMipMaps () {
+		if (!hasMipmaps) {
+			bind();
+			Gdx.gl20.glGenerateMipmap(glTarget);
```

Comment:
`hasMipmaps = true` has to be set by the implementation, since it is (currently) set by the constructor & thus requires a concrete implementation override IMO.

https://github.com/libgdx/libgdx/pull/3459/files#diff-0afdb6d047d46bcb42d807105d7084d4R134

Unless there's something else wrong with it, I'd say it's as OK as I can make it without changing current code.